core/src/main/java/io/questdb/cairo/CairoEngine.java

```diff
@@ -901,6 +916,282 @@ public TableToken lockTableName(CharSequence tableName, String dirName, int tabl
+     * for example, for testing purposes.
+     */
+    public void metadataCacheHydrateAllTables() {
+        ObjHashSet<TableToken> tableTokensSet = tlTokens.get();
+        getTableTokens(tableTokensSet, false);
+        ObjList<TableToken> tableTokens = tableTokensSet.getList();
+        TableToken tableToken = tableTokens.getQuick(0);
+
+        try {
+            for (int i = 0, n = tableTokens.size(); i < n; i++) {
+                tableToken = tableTokens.getQuick(i);
+                if (!tableToken.isSystem()) {
+                    metadataCacheHydrateTable(tableTokens.getQuick(i), true, true);
+                }
+            }
+        } catch (CairoException ex) {
+            LOG.error().$("could not hydrate metadata: [table=").$(tableToken).I$();
+        } finally {
+            tlTokens.get().clear();
+        }
+    }
+
+    /**
+     * Hydrates metadata cache for a table, using thread-local Path/ColumnVersionReader objects.
+     *
+     * @param token
+     * @param blindUpsert
+     * @see CairoEngine#metadataCacheHydrateTable(TableToken, Path, ColumnVersionReader, boolean, boolean)
+     */
+    public void metadataCacheHydrateTable(@NotNull TableToken token, boolean blindUpsert, boolean infoLog) {
+        LOG.debug().$("hydrating table using thread-local path and column version reader [table=").$(token.getTableName()).I$();
+        try {
+            metadataCacheHydrateTable(token, tlPath.get(), tlColumnVersionReader.get(), blindUpsert, infoLog);
+        } finally {
+            tlPath.get().close();
+            tlColumnVersionReader.get().close();
+        }
+    }
+
+    /**
+     * Hydrates table metadata, bypassing TableWriter/Reader.
+     * <p>
+     * This function reads the table metadata from file directly, bypassing TableReader/Writer. This ensures that it is
+     * non-blocking.
+     * <p>
+     * One must be careful on the setting of the `blindUpsert` value. When set to true, the data will be blindly
+     * upserted to the tables list, which could clobber any concurrent metadata update, leading to inconsistent state.
+     * <p>
+     * In general, any metadata change that does not originate from TableWriter (or friends) should use `blindUpsert=false`.
+     *
+     * @param token               The table token for the table to read metadata.
+     * @param path                Specify a Path object to be used.
+     * @param columnVersionReader Specify a ColumnVersionReader object to be used.
+     * @param blindUpsert         Specifies whether upsert is blind or only if non-null. This is important as TableWriter should take priority over other processes calling this function (async hydration job, queries)
+     */
+    public void metadataCacheHydrateTable(@NotNull TableToken token, @NotNull Path path, @NotNull ColumnVersionReader columnVersionReader,
+                                          boolean blindUpsert, boolean infoLog) {
+        if (infoLog) {
```

Comment:
this method is OK to call on startup, however all other metadata mutations come from TableWriter. For that we need a new method that uses TableWriter metadata instead of reading disk. The rationale is that:

- it reduced method complexity at the runtime
- there is no need to be generating new strings for table name and columns, those are already strings on the writer metadata